app/actions/websocket/roles.ts

```diff
@@ -0,0 +1,111 @@
+import {prepareMyTeams} from '@queries/servers/team';
+import {prepareUsers} from '@queries/servers/user';
+import {WebSocketMessage} from '@typings/api/websocket';
+
+import type {Model} from '@nozbe/watermelondb';
+import type {ServerDatabase} from '@typings/database/database';
+import type RoleModel from '@typings/database/models/servers/role';
+
+export async function handleRoleUpdatedEvent(serverUrl: string, msg: WebSocketMessage): Promise<void> {
+    const database = DatabaseManager.serverDatabases[serverUrl];
+    if (!database) {
+        return;
+    }
+
+    // only update Role records that exist in the Role Table
+    const role = JSON.parse(msg.data.role);
+    const dbRole = await queryRoleById(database.database, role.id);
+    if (!dbRole) {
+        return;
+    }
+
+    await database.database.write(async () => {
+        await dbRole.update((roleRecord: RoleModel) => {
+            roleRecord.permissions = role.permissions;
+        });
+    });
+}
+
+export async function handleUserRoleUpdatedEvent(serverUrl: string, msg: WebSocketMessage): Promise<void> {
+    const database = DatabaseManager.serverDatabases[serverUrl];
+    if (!database) {
+        return;
+    }
+
+    const currentUserId = await queryCurrentUserId(database.database);
+    if (currentUserId !== msg.data.user_id) {
+        return;
+    }
+
+    const modelPromises: Array<Promise<Model[]>> = [];
+
+    // update Role Table if needed
+    const newRoles = await fetchRolesIfNeeded(serverUrl, Array.from(msg.data.roles), true);
+    const preparedRoleModels = getPreparedRoleModels(database, newRoles);
+    modelPromises.push(preparedRoleModels);
+
+    // update User Table record
+    const meData = await fetchMe(serverUrl, true);
+    const userModels = prepareUsers(database.operator, [meData.user!]);
+    if (userModels) {
+        modelPromises.push(userModels);
+    }
+
+    const models = await Promise.all(modelPromises);
+    const flattenedModels = models.flat() as Model[];
+    if (flattenedModels?.length > 0) {
+        await database.operator.batchRecords(flattenedModels);
+    }
+}
+
+export async function handleMemberRoleUpdatedEvent(serverUrl: string, msg: WebSocketMessage): Promise<void> {
+    const database = DatabaseManager.serverDatabases[serverUrl];
+    if (!database) {
+        return;
+    }
+
+    const currentUserId = await queryCurrentUserId(database.database);
+    const member = JSON.parse(msg.data.member);
+    if (currentUserId !== member.user_id) {
+        return;
+    }
+
+    const modelPromises: Array<Promise<Model[]>> = [];
+
+    // update Role Table if needed
+    const newRoles = await fetchRolesIfNeeded(serverUrl, Array.from(member.roles), true);
+    const preparedRoleModels = getPreparedRoleModels(database, newRoles);
+    modelPromises.push(preparedRoleModels);
+
+    // update MyTeam Table
+    const preparedMyTeam = prepareMyTeams(database.operator, [], [member]);
```

Comment:
WS message contains the TeamMembership. so we don't need to fetch it from the server. Just run it through the `prepareMyTeams` function.

Reply:
I was currious about the `MyChannel` table `roles` field getting updated.  Looking at the server there are `channel_member_updated` and `channel_scheme_updated` WS events for updating this table so it will be addressed in a different PR.